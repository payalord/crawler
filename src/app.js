import express from 'express'
import Crawler from './crawler'
import { parse } from 'node-html-parser'

const app = express()

Array.prototype.FilterSameOrigin = function(origin) {
    return this.filter((item) => {
        const href = item.getAttribute('href')
        if (href.startsWith('/') || href.startsWith('#') || href.startsWith(`http://${origin}`) || href.startsWith(`https://${origin}`)) {
            return true
        }
        return false
    })
}

Array.prototype.FilterExternalOrigin = function(origin) {
    return this.filter((item) => {
        const href = item.getAttribute('href')
        if (href.startsWith('/') || href.startsWith('#') || href.startsWith(`http://${origin}`) || href.startsWith(`https://${origin}`)) {
            return false
        }
        return true
    })
}

String.prototype.rmQueryString = function() {
    return this.split(/[?#]/)[0]
}

const fetchImagesLinks = (root) => {
    return root.querySelectorAll('img').map(i => i.getAttribute('src'))
}

const Action = async (origin, url, processedLinks, externalProcessedLinks) => {
    const data = await Crawler((url && url.startsWith('http') && url) || (url && `https://${origin}/${url}/`) || origin)
    if (data) {
        const root = parse(data);
        const elements = root.querySelectorAll('a')
        
        const externalLinks = [...new Set(elements?.FilterExternalOrigin(origin))]
        externalLinks.map(i => {
            const href = i.getAttribute('href').rmQueryString()
            if (externalProcessedLinks.findIndex(l => l.element.getAttribute('href').rmQueryString() === href || l.element.getAttribute('href').rmQueryString()+'/' === href) === -1) {
                externalProcessedLinks.push({ element: i })
            }
        })

        const links = [...new Set(elements?.FilterSameOrigin(origin))]

        await Promise.all(
            links.map(async i => {
                const href = i.getAttribute('href').rmQueryString()
                
                if (processedLinks.findIndex(l => l.element.getAttribute('href').rmQueryString() === href || l.element.getAttribute('href').rmQueryString()+'/' === href) === -1) {
                    processedLinks.push({ element: i, images: fetchImagesLinks(root)})
                    await Action(origin, href, processedLinks, externalProcessedLinks)
                }
            })
        )
    }
}

app.get('/', async (req, res, next) => {
    if (req.query.origin) {
        const processedLinks = []
        const externalProcessedLinks = []
        console.log('Starting to crawl, please wait...')
        await Action(req.query.origin, null, processedLinks, externalProcessedLinks)
        processedLinks.push(...externalProcessedLinks)
        console.log('Full crawl completed')
        res.json(processedLinks.map(i => {return {location: i.element.getAttribute('href'), images: i.images}}))
        next()
    } else {
        res.json([])
        next()
    }
})

export default app