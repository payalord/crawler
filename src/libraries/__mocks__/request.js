// __mocks__/request.js
const mock = [{
    url: 'https://localhost:3000/',
    html: `<!DOCTYPE html>
    <html>
    <body>
    <a href="https://localhost:3000/test1">Test Page 1</a>
    <a href="https://localhost:3000/test2">Test Page 2</a>
    <a href="https://example.com/test3">Example of External link</a>
    <img src="https://localhost:3000/image1.png" />
    <img src="https://localhost:3000/image2.png" />
    </body>
    </html>`
}, {
    url: 'https://localhost:3000/test1',
    html: `<!DOCTYPE html>
    <html>
    <body>
    <a href="https://localhost:3000/test2">Test Page 2</a>
    <a href="https://example.com/test3">Example of External link</a>
    <img src="https://localhost:3000/image1.png" />
    <img src="https://localhost:3000/image2.png" />
    </body>
    </html>`
}, {
    url: 'https://localhost:3000/test2',
    html: `<!DOCTYPE html>
    <html>
    <body>
    <a href="https://localhost:3000/test1">Test Page 1</a>
    <a href="https://example.com/test3">Example of External link</a>
    <img src="https://localhost:3000/image1.png" />
    <img src="https://localhost:3000/image2.png" />
    </body>
    </html>`
}]
  
export default function request(options) {
    return new Promise((resolve, reject) => {
        const url = options.url.split('?')[0]
        const item = mock.find(i => i.url === url)
        process.nextTick(() =>
            item
                ? resolve(item.html)
                : reject({
                    error: 'Page not found.',
                })
        );
    });
}