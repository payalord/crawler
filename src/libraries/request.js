import fetch from 'isomorphic-unfetch';

const request = function(options) {
    const defaultFetchOptions = {
        method: 'GET'
    };
    const defaultHeaders = {
        Accept: 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8'
    };

    // Options
    const fetchOptions = Object.assign(defaultFetchOptions, options);
    fetchOptions.headers = Object.assign(defaultHeaders, fetchOptions.headers);

    // Add base url
    let url = fetchOptions.url;
    delete fetchOptions.url;

    // On success
    const success = function (response) {
        return response.text();
    }

    // On error
    const error = function (error) {
        console.error('Request Failed:', error);
        return Promise.reject(error);
    }

    return fetch(url, fetchOptions).then(success).catch(error);
}

export default request;
