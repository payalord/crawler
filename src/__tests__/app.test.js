import request from 'supertest'
import app from '../app'
jest.mock('../libraries/request')

describe('GET /', () => {
    it('status should be 200 with no query parameters', done => {
        request(app)
            .get('/')
            .expect(200)
            .end(function (err, res) {
                if (err) return done(err)
                expect(res.text).toMatch('[]')
                done()
            })
    })

    it('status should be 200', done => {
        request(app)
            .get('/?origin=localhost:3000')
            .expect(200)
            .end(function (err, res) {
                if (err) return done(err)
                expect(res.text).toMatchSnapshot()
                done()
            })
    })
})