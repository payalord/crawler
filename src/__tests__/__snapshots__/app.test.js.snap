// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GET / status should be 200 1`] = `"[{\\"location\\":\\"https://localhost:3000/test1\\",\\"images\\":[\\"https://localhost:3000/image1.png\\",\\"https://localhost:3000/image2.png\\"]},{\\"location\\":\\"https://localhost:3000/test2\\",\\"images\\":[\\"https://localhost:3000/image1.png\\",\\"https://localhost:3000/image2.png\\"]},{\\"location\\":\\"https://example.com/test3\\"}]"`;
